lib/pbench/cli/server/report.py

```diff
@@ -0,0 +1,403 @@
+from collections import defaultdict
+import datetime
+import re
+from typing import Optional
+
+import click
+import humanize
+from sqlalchemy import inspect, select, text
+
+from pbench.cli import pass_cli_context
+from pbench.cli.server import config_setup
+from pbench.cli.server.options import common_options
+from pbench.client.types import Dataset
+from pbench.common.logger import get_pbench_logger
+from pbench.server import BadConfig
+from pbench.server.cache_manager import CacheManager
+from pbench.server.database.database import Database
+from pbench.server.database.models.datasets import Metadata
+from pbench.server.database.models.index_map import IndexMap
+
+
+class Detail:
+    """Encapsulate generation of additional diagnostics"""
+
+    def __init__(self, detail: bool):
+        self.detail = detail
+
+    def __bool__(self) -> bool:
+        return self.detail
+
+    def write(self, message: str):
+        if self.detail:
+            click.echo(f"|| {message}")
+
+
+class Verify:
+    """Encapsulate -v status messages."""
+
+    def __init__(self, verify: bool):
+        self.verify = verify
+
+    def __bool__(self) -> bool:
+        return self.verify
+
+    def status(self, message: str):
+        if self.verify:
+            ts = datetime.datetime.now()
+            click.echo(f"({ts:%H:%M:%S}) {message}")
+
+
+class Watch:
+    """Encapsulate a periodic status check.
+
+    Discovery (especially for cache and backup) can take a long time, so we
+    centralize a periodic update notice mechanism.
+    """
+
+    def __init__(self, interval: float):
+        self.interval = datetime.timedelta(seconds=interval) if interval else None
+        self.start = datetime.datetime.now()
+        self.last = self.start
+
+    def update(self, status: str):
+        now = datetime.datetime.now()
+        if self.interval and now > self.last + self.interval:
+            self.last = now
```

Comment:
Honestly I'd considered re-casting this in terms of a strict `time()` value rather than creating, adding, and comparing `datetime` values just because the report generator (the `--cache` sweep in particular) is extremely slow and knocking even a few milliseconds off each of the hundred and seventeen thousand iterations would help. This didn't quite hit my patience threshold at the time, but maybe I should ...